awslambda/src/main/scala/akka/stream/alpakka/awslambda/AwsLambdaFlowStage.scala

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.awslambda
+
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import akka.stream.{Attributes, FlowShape, Inlet, Outlet}
+import com.amazonaws.handlers.AsyncHandler
+import com.amazonaws.services.lambda.AWSLambdaAsyncClient
+import com.amazonaws.services.lambda.model.{InvokeRequest, InvokeResult}
+
+import scala.util.control.NonFatal
+
+final class AwsLambdaFlowStage(awsLambdaClient: AWSLambdaAsyncClient)(parallelism: Int)
+  extends GraphStage[FlowShape[InvokeRequest, InvokeResult]] {
+
+  val in = Inlet[InvokeRequest]("AwsLambda.in")
+  val out = Outlet[InvokeResult]("AwsLambda.out")
+
+  @scala.throws[Exception](classOf[Exception])
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+
+      private var inFlight = 0
+
+      def invokeComplete(result: InvokeResult): Unit = {
+        inFlight -= 1
+        if (isAvailable(out)) {
+          if (!hasBeenPulled(in)) tryPull(in)
+          push(out, result)
+        }
+      }
+
+      private val asyncCallback = getAsyncCallback(invokeComplete)
+
+      private val asyncHandler = new AsyncHandler[InvokeRequest, InvokeResult] {
+        override def onError(exception: Exception): Unit =
+          exception match {
+            case NonFatal(ex) => failStage(ex)
```

Comment:
Are you sure that `onError` will only be invoked on the caller's thread, i.e. of the thread of the stage? otherwise `failStage` must be called via an `asyncCallback`.